### PR TITLE
Profile picture upload - Fixes #1263

### LINKED ISSUE #1263: Posibilidad de modificar la foto de perfil
Dado que los login sociales son muy poco usados en organizaciones privadas, las fotos de perfil son prácticamente inexistentes. Sería bueno poder cambiarla.

### PR DESCRIPTION
**Pull request by rodri042**
Fixes #1263
Depends on https://github.com/mumuki/mumuki-domain/pull/30
Demo: https://youtu.be/RqAqyYI1RlA, overlay update: https://youtu.be/pLtiQAbfA28

### CODE REVIEW

File: app/assets/stylesheets/application/_user_profile.scss [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  color: #ff5b81;
+  cursor: pointer
+}
+
+.mu-avatar-button:hover {
+  color: #ff285a;
+}
+
+.mu-avatar-button:active {
+  color: #ff043e;
```
Comment (rodri042) [inline app/assets/stylesheets/application/_user_profile.scss]: Do we have these shades of pink as SCSS variables? I couldn't find them

File: app/assets/stylesheets/application/_user_profile.scss [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+.mu-avatar-field {
+  width: 0.1px;
+  height: 0.1px;
+  opacity: 0;
+  overflow: hidden;
+  position: absolute;
+  z-index: -1;
+}
+
```
Comment (flbulgarelli) [inline app/assets/stylesheets/application/_user_profile.scss]: @luchotc @cucusita @fedescarpa I would like to get your comments about the proposed UI.

File: spec/dummy/db/schema.rb [hunk 3]
```diff
@@ -116,6 +137,7 @@
     t.integer "user_id"
     t.boolean "started", default: false
     t.datetime "started_at"
+    t.string "session_id"
```
Comment (flbulgarelli) [inline spec/dummy/db/schema.rb]: This should not be here. You may have an old migration.

File: mumuki-laboratory.gemspec [hunk 4]
```diff
@@ -16,9 +16,9 @@ Gem::Specification.new do |s|
 
   s.test_files = Dir["spec/**/*"]
 
-  s.add_dependency "rails", "~> 5.1.6"
+  s.add_dependency "rails", "~> 5.2"
 
-  s.add_dependency 'mumuki-domain', '~> 6.6.1'
+  s.add_dependency 'mumuki-domain', '~> 99.99.99' # TODO: Update gem
```
Comment (flbulgarelli) [inline mumuki-laboratory.gemspec]: Instead of doing this, you can place a `gem github: ...., ref: ...` directive in the Gemfile so that you can work even against of an unreleased version. Then we remove those directives when releasing new versions. 

That way travis can pass and we can merge the PR even before we have a release candidate for the dependency. 

This is specially true when dealing with changes in domain, since we usually release domain, bibliotheca and laboratory altogether, using the same major and minor version.

File: config/routes.rb [hunk 5]
```diff
@@ -101,6 +100,8 @@
     end
   end
 
-  #Rescue not found routes
-  get '*not_found', to: 'application#not_found'
+  # Rescue not found routes
+  get '*not_found', to: 'application#not_found', constraints: -> (req) do
+    !(req.fullpath =~ /^\/rails\/active_storage\/.*/)
```
Comment (flbulgarelli) [inline config/routes.rb]: Why is this necessary?

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -8,6 +8,11 @@ def show
   end
 
   def update
+    if avatar.present?
+      current_user.avatar.attach avatar
+      current_user.image_url = url_for(current_user.avatar)
```
Comment (flbulgarelli) [inline app/controllers/users_controller.rb]: How does this deal with current images fetched from social providers? Can both types of images coexist?

File: app/assets/stylesheets/application/_user_profile.scss (hunk 2 above)
Comment (luchotc) [reply app/assets/stylesheets/application/_user_profile.scss]: I would prefer that the upload is triggered by clicking in the picture.
Something like this would be great: 
Screenshot from 2019-04-17 11-35-02

Comment (rodri042) [reply app/assets/stylesheets/application/_user_profile.scss]: @luchotc OK, I'll try to match that design

File: mumuki-laboratory.gemspec (hunk 4 above)
Comment (rodri042) [reply mumuki-laboratory.gemspec]: That's great to know! I'll use it

File: config/routes.rb (hunk 5 above)
Comment (rodri042) [reply config/routes.rb]: That's the route where the uploads can be accessed, it's a wrapper that uses disk storage on development and the configured provider on production (an S3 bucket for example)

This "not found" route was preventing the ActiveStorage controller routes to work, since the engine adds them at the very end of the list.

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (rodri042) [reply app/controllers/users_controller.rb]: When a user changes its image, the one from the social provider is overwritten and get lost forever.

Maybe we can add a `default_image_url` with the social provider image, and a button in the UI to "reset" the avatar to the default one (in case there's any)

**Comment by rodri042** [comment]
> What we we do - with the lesser amount of work - to fix that?

Some ideas:
- Show a warning after selecting a file when the image's aspect ratio is not a perfect square
- Implement a trim library like [Croppie](https://foliotek.github.io/Croppie/)

Anyway, we should always resize the image to 250x250 on server-side, I'll look into https://edgeguides.rubyonrails.org/active_storage_overview.html#transforming-images

File: app/assets/stylesheets/application/_user_profile.scss (hunk 2 above)
Comment (rodri042) [reply app/assets/stylesheets/application/_user_profile.scss]: Done! https://youtu.be/pLtiQAbfA28

File: app/assets/stylesheets/application/_user_profile.scss (hunk 1 above)
Comment (rodri042) [reply app/assets/stylesheets/application/_user_profile.scss]: This button doesn't exist anymore

**Comment by luchotc** [comment]
@rodri042 Beautiful!

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (luchotc) [reply app/controllers/users_controller.rb]: I think we are just missing this and we are ready to go!

**Comment by luchotc** [comment]
We've finally decided to use Mumuki provided avatars instead of allowing users to upload their own, so I'm closing this one :cry:
